### ISSUE TITLE
Weird memory issue with git_revwalk_next. (Only seeing this issue on macOS.)

### ISSUE THREAD

**Issue by nkulevsky** [libgit2/libgit2#6582]
Hello -- I'm seeing some behavior that has me questioning my sanity.

I've recreated it by making a minimal C++ program that seemingly recreates it.

If, using any fairly large repo, I iterate on it using `git_revwalk_next` as in the snippet below, I seemingly am encountering a memory leak, but only on macOS, not on Windows or Linux.

If I run my example on a large repo, when the loop is done, (and I've free'd everything), my system tools on macOS (Activity Manager, ps, top, etc) show I'm still consuming 20 MB or so of resident memory.

On windows or Linux, the memory seems to be cleaned up by the time I'm done iterating and have free'd everything.

I've run Xcode's Instruments profiler against this, and it shows no leaks, and it also shows the ~20 MB of memory being released when everything is free'd, as I would expect.

Are macOS's system tools, such as Activity Manager, ps top, etc lying to me? XCode's instruments seems happy with things, and I'm not sure what to believe.

### Reproduction steps

I created this minimal program to repro:

```c++
#include "git2.h"
#include <iostream>

int main()
{

    std::cout << "Starting!" << std::endl;

    char input;
    std::cin>>input;

    git_libgit2_init();

    git_repository *repo = nullptr;
    git_repository_open(&repo, "[any valid repo with a good amount of history]");

    git_revwalk *walker = nullptr;
    git_revwalk_new(&walker, repo);
    git_revwalk_push_head(walker);
    git_oid oid;

    std::cout << "Iterating repo" << std::endl;

    int iter_count = 0;
    while (git_revwalk_next(&oid, walker) != GIT_ITEROVER && iter_count++ < 50000)
    {
        std::cout << "Iteration: " << iter_count << std::endl;
    }
    
    git_revwalk_reset(walker);
    git_revwalk_free(walker);
    git_repository_free(repo);

    git_libgit2_shutdown(); 
 
    std::cin>>input;

}
``` 

### Expected behavior

I would expect to see the ~20 MB consumed by libgit2 during my iterations to be returned to me after I've free'd and shutdown the libgit2 lib. 

### Actual behavior
Activity Monitor, ps, and top on macOS are telling me that I'm still consuming ~20MB of resident memory. Running this on Linux or Windows behaves as expected.

### Version of libgit2 (release number or SHA1)

1.6.4

### Operating system(s) tested
macOS Monterey (12.6.6)

**Comment by neithernut** [comment]
Could you try running your reproducer under valgrind (specifically `memcheck`)?

**Comment by nkulevsky** [comment]
Thanks. I have done so.

I ran valgrind like this:

```shell
valgrind --leak-check=full --track-origins=yes --log-file=valgrind-out.txt ./harness/runner
``` 

(harness/runner is the executable with the code listed above.)

As a sanity check, I added this one line at the end of my code, on line 36, simply to ensure it caught this obvious memory leak:

```c++
    __unused auto x = malloc(10000000);
``` 

Anyway, I ran with valgrind, and it seemed to complain about some invalid reads rooted in libGit, but no obvious memory issues. It does complain about a few bytes deep in some system/valgrind files, so not sure what means, but it doesn't seems to blame anything on libGit. 

It does call out the obvious mem leak I mentioned above, just so I could verify it was otherwise functioning.

I guess at this point, I'm kind of at a loss. Both valgrind and xCode's Instruments tool don't see anything, but Activity Manager/top/PS seem to see resident memory allocated.

[valgrind-out.txt](https://github.com/libgit2/libgit2/files/11834989/valgrind-out.txt)
